graph/src/components/store/entity_cache.rs

```diff
@@ -203,18 +202,111 @@ impl EntityCache {
+                // mark the key for removal from the map
+                keys_to_remove.push(key.clone());
+            }
+        }
+
+        // A helper function that checks if an update matches the query and returns the updated entity if it does
+        fn matches_query(
+            op: &EntityOp,
+            query: &DerivedEntityQuery,
+            key: &EntityKey,
+        ) -> Result<Option<Entity>, anyhow::Error> {
+            match op {
+                EntityOp::Update(entity) | EntityOp::Overwrite(entity)
+                    if query.matches(key, entity) =>
+                {
+                    Ok(Some(entity.clone()))
+                }
+                EntityOp::Remove => Ok(None),
+                _ => Ok(None),
+            }
+        }
+
+        // Iterate over self.updates to find entities that:
+        // - Aren't already present in the entity_map
+        // - Match the query
+        // If these conditions are met:
+        // - Check if there's an update for the same entity in handler_updates and apply it.
+        // - Add the entity to entity_map.
+        for (key, op) in self.updates.iter() {
+            if !entity_map.contains_key(key) {
+                if let Some(entity) = matches_query(op, &query, key)? {
+                    if let Some(handler_op) = self.handler_updates.get(key).cloned() {
+                        // If there's a corresponding update in handler_updates, apply it to the entity
+                        // and insert the updated entity into entity_map
+                        let mut entity_cow = Some(Cow::Borrowed(&entity));
+                        handler_op
+                            .apply_to(&mut entity_cow)
+                            .map_err(|e| key.unknown_attribute(e))?;
+
+                        if let Some(updated_entity) = entity_cow {
+                            entity_map.insert(key.clone(), updated_entity.into_owned());
+                        }
+                    } else {
+                        // If there isn't a corresponding update in handler_updates or the update doesn't match the query, just insert the entity from self.updates
+                        entity_map.insert(key.clone(), entity);
+                    }
+                }
+            }
+        }
+
+        // Iterate over handler_updates to find entities that:
+        // - Aren't already present in the entity_map.
+        // - Aren't present in self.updates.
+        // - Match the query.
+        // If these conditions are met, add the entity to entity_map.
+        for (key, handler_op) in self.handler_updates.iter() {
+            if !entity_map.contains_key(key) && !self.updates.contains_key(key) {
+                if let Some(entity) = matches_query(handler_op, &query, key)? {
+                    entity_map.insert(key.clone(), entity);
+                }
+            }
+        }
+
+        // Remove keys that were marked for removal
+        for key in keys_to_remove {
```

Comment:
The removal is placed at last because for the stuff that has been removed we don't need to reprocess it in the `updates` and `handler_updates` section (which only care about stuff that is not in `entity_map`. So if the removal is immediatly after the processing entities from store `updates` section will reapply the changes unneccessarily. The alternate is to convert `keys_to_remove` a `HashSet` and have another condition `!keys_to_remove.contains()` in the `updates` and `handler_updates` section when iterating over them

Reply:
I agree with the reasoning; it would be good to capture some of that in the comment, something like
```
// Remove entities that are in the store but have been removed by an update. 
// We do this last since the loops over updates and handler_updates are only
// concerned with entities that are not in the store yet and by leaving removed 
// keys in entity_map we avoid processing these updates a second time when we
// already looked at them when we went through entity_map
```